File: backend/app/services/profile_service.py
from sqlalchemy.orm import Session

from app.models.user import User
from app.services.nutrition_goal_service import calculate_daily_goals


def calculate_bmi(weight, height):
    height_m = height / 100
    bmi = round(weight / (height_m ** 2), 2)

    if bmi < 18.5:
        category = "Underweight"
    elif bmi < 25:
        category = "Normal"
    elif bmi < 30:
        category = "Overweight"
    else:
        category = "Obese"

    return bmi, category


def update_profile(
    user_id: int,
    age: int,
    gender: str,
    height: float,
    weight: float,
    goal: str,
    activity_level: str,
    health_condition: str,
    db: Session,
):

    user = db.query(User).filter(User.id == user_id).first()

    if not user:
        return None

    # -----------------------------
    # Update User Profile
    # -----------------------------
    user.age = age
    user.gender = gender
    user.height = height
    user.weight = weight
    user.goal = goal
    user.activity_level = activity_level
    user.health_condition = health_condition

    # -----------------------------
    # Calculate BMI
    # -----------------------------
    bmi, category = calculate_bmi(weight, height)

    user.bmi = bmi
    user.bmi_category = category

    # -----------------------------
    # Calculate Daily Nutrition Goals
    # -----------------------------
    goals = calculate_daily_goals(user)

    print("\n========== DAILY GOALS ==========")
    print(goals)

    # -----------------------------
    # Save Daily Goals
    # -----------------------------
    user.daily_calories = goals["calories"]
    user.daily_protein = goals["protein"]
    user.daily_carbs = goals["carbs"]
    user.daily_fat = goals["fat"]
    user.daily_water = goals["water"]

    print("\n========== BEFORE COMMIT ==========")
    print("Calories :", user.daily_calories)
    print("Protein  :", user.daily_protein)
    print("Carbs    :", user.daily_carbs)
    print("Fat      :", user.daily_fat)
    print("Water    :", user.daily_water)

    db.commit()
    db.refresh(user)

    print("\n========== AFTER COMMIT ==========")
    print("Calories :", user.daily_calories)
    print("Protein  :", user.daily_protein)
    print("Carbs    :", user.daily_carbs)
    print("Fat      :", user.daily_fat)
    print("Water    :", user.daily_water)

    return user


def get_profile(user_id: int, db: Session):
    return db.query(User).filter(User.id == user_id).first()

File: frontend/src/services/profileService.ts
import { apiClient } from "./apiClient";
import type {
  ProfileUpdatePayload,
  UserProfile,
} from "../types/profile";

import {
  getCachedUserProfile,
  getStoredAuthUser,
  saveUserProfile,
} from "../utils/profileStorage";

const normalizeProfile = (
  data: UserProfile,
  cached?: UserProfile | null
): UserProfile => ({
  full_name: data.full_name ?? cached?.full_name ?? "",
  email: data.email ?? cached?.email ?? "",

  age: data.age ?? cached?.age ?? null,
  gender: data.gender ?? cached?.gender ?? null,
  height: data.height ?? cached?.height ?? null,
  weight: data.weight ?? cached?.weight ?? null,

  goal: data.goal ?? cached?.goal ?? null,
  activity_level:
    data.activity_level ?? cached?.activity_level ?? null,

  health_condition:
    data.health_condition ?? cached?.health_condition ?? null,

  bmi: data.bmi ?? cached?.bmi ?? null,
  bmi_category:
    data.bmi_category ?? cached?.bmi_category ?? null,

  // Daily Nutrition Goals
  daily_calories:
    data.daily_calories ?? cached?.daily_calories ?? 0,

  daily_protein:
    data.daily_protein ?? cached?.daily_protein ?? 0,

  daily_carbs:
    data.daily_carbs ?? cached?.daily_carbs ?? 0,

  daily_fat:
    data.daily_fat ?? cached?.daily_fat ?? 0,

  daily_water:
    data.daily_water ?? cached?.daily_water ?? 0,

  member_since:
    data.member_since ?? cached?.member_since ?? null,

  avatar_uri:
    data.avatar_uri ?? cached?.avatar_uri ?? null,
});

export const getProfile = async (): Promise<UserProfile> => {
  const authUser = await getStoredAuthUser();

  if (!authUser?.id) {
    throw new Error("User not authenticated");
  }

  try {
    const response = await apiClient.get<UserProfile>("/profile");

    const cached = await getCachedUserProfile(authUser.id);

    const profile = normalizeProfile(
      response.data,
      cached
    );

    await saveUserProfile(authUser.id, profile);

    return profile;
  } catch (error) {
    const cached = await getCachedUserProfile(authUser.id);

    if (cached) {
      return cached;
    }

    throw error;
  }
};

export const updateProfile = async (
  profileData: ProfileUpdatePayload
): Promise<UserProfile> => {
  const authUser = await getStoredAuthUser();

  if (!authUser?.id) {
    throw new Error("User not authenticated");
  }

  const response = await apiClient.put<UserProfile>(
    "/profile",
    profileData
  );

  const cached = await getCachedUserProfile(authUser.id);

  const profile = normalizeProfile(
    response.data,
    cached
  );

  await saveUserProfile(authUser.id, profile);

  return profile;
};

export const getCachedProfile = async (): Promise<UserProfile | null> => {
  const authUser = await getStoredAuthUser();

  if (!authUser?.id) {
    return null;
  }

  return getCachedUserProfile(authUser.id);
};